src/static/js/collab_client.js

```diff
@@ -182,6 +163,26 @@ const getCollabClient = (ace2editor, serverVars, initialUserInfo, options, _pad)
         });
   };
 
+  const serverMessageTaskQueue = new class {
+    constructor() {
+      this._q = Promise.resolve();
+    }
+
+    async enqueue(fn) {
+      const p = this._q.then(async () => {
+        while (true) {
+          const composing = editor.getInInternationalComposition();
+          if (!composing) break;
+          await composing;
+        }
+        await fn(); // It's OK if fn() doesn't return a Promise.
```

Comment:
If I were to simplify it, I would go a step further:

```javascript
await editor.getInInternationalComposition();
await fn(); // It's OK if fn() doesn't return a Promise.
```

This works because `await null;` is equivalent to `await Promise.resolve(null);`.

The reason I didn't do that is because I was worried there might still be a race condition: Can a new `compositionstart` event fire between those two lines of code? If I understand correctly, the ECMAScript standard says that the second line is scheduled for future execution even if the Promise returned in the first line is already resolved (or null). Maybe it's not possible for a `compositionstart` event to fire between those two lines; I don't know enough about the specs to say either way. If it is possible, maybe it's so unlikely that it would be better to have more readable code. Or, maybe it is somewhat likely, but there are no negative consequences so it doesn't matter. I erred on the side of caution at the expense of readability. 

Either way I should at least add a comment.

Reply:
I did a short simulation of a new call to `compositionstart` right after resolving the `compositionend` promise.
https://jsfiddle.net/ax0phrdg/

What I noticed is the `fn` will be called right after the first `compositionend`, which is good. I mean, a new event cycle will not lock the `getInInternationalComposition` promise, so I think we are safe regarding the race condition.

Reply:
Done.